### ISSUE TITLE
Easily port models for Unity ml-agents [feature request]

### ISSUE THREAD

**Issue by TravisDunlop** [hill-a/stable-baselines#58]
I am trying to create reinforcement learning agents for the [Unity ml-agents](https://github.com/Unity-Technologies/ml-agents) simulation environment.  While Unity ml-agents has some built in RL model trainers, I've found that stable-baselines runs faster.

I would like to be able to pass the trained policy from stable-baselines back into Unity.  To do this, Unity ml-agents requires a .bytes file.  This file is a byte code protocol buffer that has both the frozen Tensorflow graph along with the latest session checkpoint containing the TensorFlow Variable values.

The Unity ml-agents documentation for creating this .bytes file is [here](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Using-TensorFlow-Sharp-in-Unity.md#using-your-own-trained-graphs) Under the 'Using your own trained graphs' section.  In the documentation they refer to this type of .bytes file as an 'Internal Brain.'

By the way, I love stable-baselines, it's great!

**Comment by araffin** [comment]
Hi, 
Thanks for your interest =)

I have never used Unity ml-agents, but could be a good idea to enable integration with it.

>I would like to be able to pass the trained policy from stable-baselines back into Unity. To do this, >Unity ml-agents requires a .bytes file.
Is there no other way, it does not follow Gym interface? A better solution would be to do a "gym-unity-bridge" no?

**Comment by TravisDunlop** [comment]
Hi Antonin, 

I ended up putting together a utility to freeze the graph to a protocol buffer.  This turns all variables to constants.  So, it's really just used when ready to serve the model, both in Unity and other applications.  I borrowed heavily from this excellent [blog series](https://blog.metaflow.fr/tensorflow-how-to-freeze-a-model-and-serve-it-with-a-python-api-d4f3596b3adc).

Here is the [gist with code](https://gist.github.com/TravisDunlop/500577a8b491c420581b7713af98e247).

If you have a stable-baselines `model` that you wanted to save to the protocol buffer you'd call:
```save_to_pb(model.graph, 'my_model.pb')```

As an example, here's how you'd serve the model in python (it can also be used in C++, C#, or other languages)
```
graph = load_graph('my_model.pb')
_input = graph.get_tensor_by_name(input_name)
_output = graph.get_tensor_by_name(output_name)
with tf.Session(graph=graph) as sess:
    action = self.sess.run(self._output, feed_dict={_input : observation})
````

The `input_name` and `output_name` can be found at `model.act_model.obs_ph.name` and `model.act_model.action.name` respectively.  Just make sure to remove the `:0` that is attached to the end of each of these strings.  If you are using this script for Unity ml-agents, these strings also need to be put into the "Vector Observation Node"/"Visual Observation Placeholder Name" and "Action Node Name" fields under the Internal Brain Settings.

I'm hoping to get your thoughts on if/how this should be integrated directly into stable-baselines. :)

**Comment by araffin** [comment]
Hi,
Thank you for your contribution. From what I understand, this is only useful for testing/production?
Did you test it on unity? How did you train the model then?
It could be a good feature to have, what do you think @hill-a ?

Anyway, if this utility is added to stable-baselines (I would add your script in the `common` folder), you will need to update the doc and add some tests before submitting a PR ;)

**Comment by araffin** [comment]
@TravisDunlop I think this repo may interest you ;) 
Link: https://github.com/Sohojoe/MarathonEnvsBaselines

**Comment by araffin** [comment]
I think this was similar and now solved in #329 
Feel free to re-open it if something missing.

**Comment by araffin** [comment]
Also related: https://medium.com/aureliantactics/setting-up-unity-ml-agents-with-ray-and-stable-baselines-e4e2a72dd1b4
